### PR TITLE
HTTP CONNECT Auth support

### PR DESCRIPTION
**Pull request by yashykt**
This pull request is for adding HTTP CONNECT Auth support. The working is such that if user credentials are present in the http_proxy environment variable value, then it adds the Proxy-Authorization header to the request.

### CODE REVIEW

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

**Comment by grpc-testing** [comment]
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok for us to run on our infrastructure.

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 1]
```diff
@@ -47,15 +50,18 @@ static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
...
   }
-  if (strchr(uri->authority, '@') != NULL) {
-    gpr_log(GPR_ERROR, "userinfo not supported in proxy URI");
-    goto done;
+  char *user_cred_end = strchr(uri->authority, '@');
+  if (user_cred_end != NULL) {
+    *name_to_resolve = gpr_strdup(user_cred_end + 1);
+    *user_cred_end = '\0';
+    *user_cred = gpr_strdup(uri->authority);
+    gpr_log(GPR_INFO, "userinfo found in proxy URI");
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/http_proxy.c]: nit: it usually helps to print the data in question, unless it's sensitive and/or too long to print.

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 2]
```diff
@@ -47,15 +50,18 @@ static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
     gpr_log(GPR_ERROR, "'%s' scheme not supported in proxy URI", uri->scheme);
     goto done;
   }
-  if (strchr(uri->authority, '@') != NULL) {
-    gpr_log(GPR_ERROR, "userinfo not supported in proxy URI");
-    goto done;
+  char *user_cred_end = strchr(uri->authority, '@');
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/http_proxy.c]: For this whole section, you may want to use `gpr_string_split` instead of doing it all manually.

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 3]
```diff
@@ -32,13 +32,16 @@
 #include "src/core/lib/channel/channel_args.h"
 #include "src/core/lib/support/env.h"
 #include "src/core/lib/support/string.h"
+#include "src/core/lib/slice/b64.h"
 
-static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
+static void grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx,
+                                char **name_to_resolve,
+                                char **user_cred) {
+  *name_to_resolve = NULL;
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/http_proxy.c]: it seems inconsistent to reset `name_to_resolve` but not `user_cred`.

File: test/core/end2end/fixtures/h2_http_proxy.c [hunk 4]
```diff
@@ -47,11 +47,23 @@ static grpc_end2end_test_fixture chttp2_create_fixture_fullstack(
...
   grpc_end2end_test_fixture f;
   memset(&f, 0, sizeof(f));
+  grpc_exec_ctx exec_ctx = GRPC_EXEC_CTX_INIT;
 
   fullstack_fixture_data *ffd = gpr_malloc(sizeof(fullstack_fixture_data));
   const int server_port = grpc_pick_unused_port_or_die();
   gpr_join_host_port(&ffd->server_addr, "localhost", server_port);
-  ffd->proxy = grpc_end2end_http_proxy_create();
+
+  //If we are testing proxy auth, add the proxy auth arg to proxy channel args
```
Comment (dgquintas) [inline test/core/end2end/fixtures/h2_http_proxy.c]: nit: space between `//` and `If`

File: test/core/end2end/fixtures/h2_http_proxy.c [hunk 5]
```diff
@@ -47,11 +47,23 @@ static grpc_end2end_test_fixture chttp2_create_fixture_fullstack(
...
+  grpc_channel_args *proxy_args = NULL;
+  const grpc_arg *proxy_auth_arg = grpc_channel_args_find(
+          client_args, GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT);
+  if(proxy_auth_arg) {
```
Comment (dgquintas) [inline test/core/end2end/fixtures/h2_http_proxy.c]: `!= NULL`

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 6]
```diff
@@ -352,6 +354,42 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
     GRPC_ERROR_UNREF(error);
     return;
   }
+  // If proxy auth is being used, check if the header is present
+  if(grpc_channel_args_find(
+      conn->proxy->channel_args,
+      GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT) != NULL) {
+    bool found = false, failed = false;
```
Comment (dgquintas) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: single statement per line.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 7]
```diff
@@ -352,6 +354,42 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+    for(size_t i = 0; i < conn->http_request.hdr_count; i++) {
+      if(strcmp(conn->http_request.hdrs[i].key, "Proxy-Authorization") == 0) {
```
Comment (dgquintas) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: consistency: you use `strcmp` here but `strncmp` below. When comparing against a static string, you needn't use `strncmp`.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 8]
```diff
@@ -352,6 +354,42 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+        found = true;
+        // Authentication type should be Basic
+        if(strncmp(conn->http_request.hdrs[i].value, "Basic",
```
Comment (dgquintas) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: s/strncmp/strcmp, and remove the `strlen("Basic")`

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 6 above)
Comment (dgquintas) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Please name these variables more descriptively: found what? what failed?

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 9]
```diff
@@ -352,6 +354,42 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+                   strlen("Basic")) != 0) {
+          failed = true;
+          break;
+        }
+        // Check if encoded string is as expected
+        char *encoded_str_start =
```
Comment (dgquintas) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Check for errors. If `strchr` doesn't find a match, there'll be dereferencing a null pointer in `grpc_base64_decode`.

File: src/core/ext/filters/client_channel/http_proxy.c (hunk 3 above)
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 10]
```diff
@@ -32,13 +32,16 @@
...
   char* uri_str = gpr_getenv("http_proxy");
-  if (uri_str == NULL) return NULL;
+  if (uri_str == NULL) return;
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done

File: src/core/ext/filters/client_channel/http_proxy.c (hunk 2 above)
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 11]
```diff
@@ -47,15 +50,18 @@ static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
     gpr_log(GPR_ERROR, "'%s' scheme not supported in proxy URI", uri->scheme);
     goto done;
   }
-  if (strchr(uri->authority, '@') != NULL) {
-    gpr_log(GPR_ERROR, "userinfo not supported in proxy URI");
-    goto done;
+  char *user_cred_end = strchr(uri->authority, '@');
+  if (user_cred_end != NULL) {
+    *name_to_resolve = gpr_strdup(user_cred_end + 1);
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: removed.

File: src/core/ext/filters/client_channel/http_proxy.c (hunk 1 above)
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: it holds the user credentials username:password. Don't want that to show up in logs

File: test/core/end2end/fixtures/h2_http_proxy.c (hunk 4 above)
Comment (yashykt) [inline test/core/end2end/fixtures/h2_http_proxy.c]: done

File: test/core/end2end/fixtures/h2_http_proxy.c (hunk 5 above)
Comment (yashykt) [inline test/core/end2end/fixtures/h2_http_proxy.c]: done

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 6 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: removed

Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done. Thanks :)

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 7 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: removed. (The reason for strncmp was that I wanted to compare only the first 'basic' portion of the header value. Changed it now to split over space)

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 8 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: removed

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 9 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Good catch! Thanks!

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 12]
```diff
@@ -34,12 +34,19 @@
...
+/**
+ * Parses the 'http_proxy' env var and returns the proxy hostname to resolve or
+ * NULL on error. Also sets 'user_cred' if it is not NULL to user credentials
+ * if present in the 'http_proxy' env var.
+ */
+static char *grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx,
                                 char **user_cred) {
-  *name_to_resolve = NULL;
+  char *proxy_name = NULL;
+  if(user_cred != NULL) {
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/http_proxy.c]: you could simply always do it. If the `if` is true, you set to NULL. Otherwise, it's already NULL, so we don't need the `if`. We can also simplify the wording of the docstring accordingly.

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 13]
```diff
@@ -50,18 +57,33 @@ static void grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx,
...
-  char *user_cred_end = strchr(uri->authority, '@');
-  if (user_cred_end != NULL) {
-    *name_to_resolve = gpr_strdup(user_cred_end + 1);
-    *user_cred_end = '\0';
-    *user_cred = gpr_strdup(uri->authority);
+  /* Split on '@' to separate user credentials from host */
+  char **authority_strs = NULL;
+  size_t authority_nstrs;
+  gpr_string_split(uri->authority, "@", &authority_strs, &authority_nstrs);
+  GPR_ASSERT(authority_nstrs != 0); /* should have atleast 1 string */
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/http_proxy.c]: s/atleast/at least

File: src/core/ext/filters/client_channel/http_proxy.c (hunk 12 above)
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: If user_cred itself is NULL, then setting *user_cred would result in a SEGFAULT

**Comment by markdroth** [comment]
Jenkins: test this please

File: test/core/end2end/fixtures/http_proxy_fixture.h [hunk 14]
```diff
@@ -16,11 +16,30 @@
  *
  */
 
+#ifndef GRPC_TEST_CORE_END2END_FIXTURES_HTTP_PROXY_FIXTURE_H
+#define GRPC_TEST_CORE_END2END_FIXTURES_HTTP_PROXY_FIXTURE_H
+
+#include <grpc/grpc.h>
+
+/* The test credentials being used for HTTP Proxy Authorization */
+#define GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED "aladdin:opensesame"
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.h]: Suggest calling this `GRPC_TEST_HTTP_PROXY_AUTH_CREDS`.

File: test/core/end2end/fixtures/http_proxy_fixture.h [hunk 15]
```diff
@@ -16,11 +16,30 @@
...
+
+/* A channel arg key used to indicate that the channel uses proxy authorization.
+ * The value is of no consequence as just the presence of the argument is
+ * enough. It is currently kept as of type integer but can be changed as seen
+ * fit.
+ */
+#define GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT \
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.h]: Suggest calling this `GRPC_ARG_HTTP_PROXY_AUTH_CREDS`.

File: test/core/end2end/fixtures/h2_http_proxy.c [hunk 16]
```diff
@@ -64,8 +77,17 @@ void chttp2_init_client_fullstack(grpc_end2end_test_fixture *f,
...
-               grpc_end2end_http_proxy_get_proxy_name(ffd->proxy));
+
+  /* If testing for proxy auth, add credentials to proxy uri */
+  if(grpc_channel_args_find(
+      client_args, GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT) == NULL) {
+    gpr_asprintf(&proxy_uri, "http://%s",
+                 grpc_end2end_http_proxy_get_proxy_name(ffd->proxy));
+  } else {
+    gpr_asprintf(&proxy_uri, "http://%s@%s",
+                 GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED,
```
Comment (markdroth) [inline test/core/end2end/fixtures/h2_http_proxy.c]: Instead of hard-coding this value, let's make the value of the channel arg be the credentials to use.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 17]
```diff
@@ -352,6 +394,30 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+            exec_ctx, conn->http_request.hdrs[i].value)) {
+          break;
+        }
+        auth_header_found = true;
+        break;
+      }
+    }
+    if(!auth_header_found) {
+      const char *msg = "HTTP Connect could not verify authentication";
+      error = GRPC_ERROR_CREATE_FROM_COPIED_STRING(msg);
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Can use `GRPC_ERROR_CREATE_FROM_STATIC_STRING()` here, since this is a string literal.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 18]
```diff
@@ -352,6 +394,30 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
     GRPC_ERROR_UNREF(error);
     return;
   }
+  // If proxy auth is being used, check if the header is present and as expected
+  if(grpc_channel_args_find(
+      conn->proxy->channel_args,
+      GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT) != NULL) {
+    bool auth_header_found = false;
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: There's probably no need for this bool.  Instead, just move the body of the `if (!auth_header_found)` block inside of the loop:

```
if (strcmp(conn->http_request.hdrs[i].key, "Proxy-Authorization") == 0) {
  if (!proxy_auth_header_matches(
          exec_ctx, conn->http_request.hdrs[i].value)) {
    // ... call proxy_connection_failed()...
    return;
  }
  break;
}
```

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 19]
```diff
@@ -304,6 +307,45 @@ static void on_server_connect_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+  }
+  // Authentication type should be Basic
+  if(strcmp(auth_header_strs[0], "Basic") != 0) {
+    goto done;
+  }
+  // should match GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED after decoding
+  grpc_slice decoded_slice =
+      grpc_base64_decode(exec_ctx, auth_header_strs[1], 0);
+  if(grpc_slice_str_cmp(
+      decoded_slice, GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED) != 0) {
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: As mentioned elsewhere, we should use the value of the channel arg instead of hard-coding this value.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 20]
```diff
@@ -304,6 +307,45 @@ static void on_server_connect_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+                   &auth_header_nstrs);
+  if(auth_header_nstrs != 2) {
+    goto done;
+  }
+  // Authentication type should be Basic
+  if(strcmp(auth_header_strs[0], "Basic") != 0) {
+    goto done;
+  }
+  // should match GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED after decoding
+  grpc_slice decoded_slice =
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: I think you need to unref this slice before returning.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 21]
```diff
@@ -304,6 +307,45 @@ static void on_server_connect_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+static bool proxy_auth_header_matches(grpc_exec_ctx *exec_ctx,
+                                      char *proxy_auth_header_val) {
+  if(proxy_auth_header_val == NULL) {
+    return false;
+  }
+  char **auth_header_strs;
+  size_t auth_header_nstrs;
+  bool auth_header_matches = false;
+  // Split the auth header value on space
+  gpr_string_split(proxy_auth_header_val, " ", &auth_header_strs,
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: In this particular case, I think `gpr_string_split()` is overkill, since we know what the first part of the string should be.  I suggest something like this instead:

```
if (strncmp(proxy_auth_header_val, "Basic ", 6) != 0) return false;
proxy_auth_header_val += 6;
gpc_slice decoded_slice =
    grpc_base64_decode(exec_ctx, proxy_auth_header_val, 0);
const bool match = grpc_slice_str_cmp(
    decoded_slice, GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED) != 0;
grpc_slice_internal_unref(exec_ctx, decoded_slice);
return match;
```

File: test/core/end2end/fixtures/http_proxy_fixture.h [hunk 22]
```diff
@@ -16,11 +16,30 @@
...
+#ifndef GRPC_TEST_CORE_END2END_FIXTURES_HTTP_PROXY_FIXTURE_H
+#define GRPC_TEST_CORE_END2END_FIXTURES_HTTP_PROXY_FIXTURE_H
+
+#include <grpc/grpc.h>
+
+/* The test credentials being used for HTTP Proxy Authorization */
+#define GRPC_END2END_HTTP_PROXY_TEST_CONNECT_CRED "aladdin:opensesame"
+
+/* A channel arg key used to indicate that the channel uses proxy authorization.
+ * The value is of no consequence as just the presence of the argument is
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.h]: As mentioned above, let's make this a string arg and have its value be the credentials to use.

File: test/core/end2end/tests/payload_with_proxy_auth.c [hunk 23]
```diff
@@ -0,0 +1,302 @@
+/*
```
Comment (markdroth) [inline test/core/end2end/tests/payload_with_proxy_auth.c]: I suggest just calling this proxy_auth.

Comment (markdroth) [inline test/core/end2end/tests/payload_with_proxy_auth.c]: It looks like you based this off of the existing payload test.  However, the simple_request test would probably be a simpler template to use, since we don't really need a message payload to test the proxy auth code.

File: test/core/end2end/tests/payload_with_proxy_auth.c [hunk 24]
```diff
@@ -0,0 +1,302 @@
...
+      config, "test_invoke_10_request_response_with_payload_and_proxy_auth",
+      &client_args, NULL);
+  for (i = 0; i < 10; i++) {
+    request_response_with_payload_and_proxy_auth(config, f);
+  }
+  end_test(&f);
+  config.tear_down_data(&f);
+}
+
+void payload_with_proxy_auth(grpc_end2end_test_config config) {
```
Comment (markdroth) [inline test/core/end2end/tests/payload_with_proxy_auth.c]: There's no point in running this test in any fixture except for h2_http_proxy_test.  So let's add a new FEATURE_MASK_* macro to indicate that the fixture uses an HTTP proxy, make sure that the h2_http_proxy_test fixture sets that bit, and then change the code here to only run the tests if that bit is set.

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 25]
```diff
@@ -32,13 +32,23 @@
...
-static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
+/**
+ * Parses the 'http_proxy' env var and returns the proxy hostname to resolve or
+ * NULL on error. Also sets 'user_cred' to user credentials present in the
+ * 'http_proxy' env var, NULL if not present.
+ */
+static char *grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx,
+                                char **user_cred) {
+  char *proxy_name = NULL;
+  if(user_cred != NULL) {
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: Done!

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 26]
```diff
@@ -47,11 +57,29 @@ static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
...
-    gpr_log(GPR_ERROR, "userinfo not supported in proxy URI");
-    goto done;
+  /* Split on '@' to separate user credentials from host */
+  char **authority_strs = NULL;
+  size_t authority_nstrs;
+  gpr_string_split(uri->authority, "@", &authority_strs, &authority_nstrs);
+  GPR_ASSERT(authority_nstrs != 0); /* should have at least 1 string */
+  if(authority_nstrs == 1) {
+  /* User cred not present in authority */
+    proxy_name = gpr_strdup(authority_strs[0]);
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done!

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 27]
```diff
@@ -47,11 +57,29 @@ static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
...
+  } else if(authority_nstrs == 2) {
+  /* User cred found */
+    if(user_cred != NULL) {
+      *user_cred = gpr_strdup(authority_strs[0]);
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done!

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 28]
```diff
@@ -47,11 +57,29 @@ static char* grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx) {
...
+    }
+    proxy_name = gpr_strdup(authority_strs[1]);
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done!

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 29]
```diff
@@ -126,10 +159,25 @@ static bool proxy_mapper_map_name(grpc_exec_ctx* exec_ctx,
...
     }
   }
-  grpc_arg new_arg = grpc_channel_arg_string_create(
+  grpc_arg args_to_add[2];
+  args_to_add[0] = grpc_channel_arg_string_create(
       GRPC_ARG_HTTP_CONNECT_SERVER,
       uri->path[0] == '/' ? uri->path + 1 : uri->path);
-  *new_args = grpc_channel_args_copy_and_add(args, &new_arg, 1);
+  if(user_cred != NULL) {
+    /* Use base64 encoding for user credentials */
```
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done!

File: test/core/end2end/fixtures/h2_http_proxy.c [hunk 30]
```diff
@@ -47,11 +47,24 @@ static grpc_end2end_test_fixture chttp2_create_fixture_fullstack(
...
   gpr_join_host_port(&ffd->server_addr, "localhost", server_port);
-  ffd->proxy = grpc_end2end_http_proxy_create();
+
+  /* If we are testing proxy auth, add the proxy auth arg to proxy channel args
+   */
+  grpc_channel_args *proxy_args = NULL;
+  const grpc_arg *proxy_auth_arg = grpc_channel_args_find(
+          client_args, GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT);
+  if(proxy_auth_arg != NULL) {
+    proxy_args = grpc_channel_args_copy_and_add(NULL, proxy_auth_arg, 1);
```
Comment (yashykt) [inline test/core/end2end/fixtures/h2_http_proxy.c]: done!

File: test/core/end2end/fixtures/h2_http_proxy.c (hunk 16 above)
Comment (yashykt) [inline test/core/end2end/fixtures/h2_http_proxy.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.h (hunk 14 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.h]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.h (hunk 22 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.h]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.h (hunk 15 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.h]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 21 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 20 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done. Thanks!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 19 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 18 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: changed

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 17 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

File: test/core/end2end/tests/payload_with_proxy_auth.c (hunk 23 above)
Comment (yashykt) [inline test/core/end2end/tests/payload_with_proxy_auth.c]: done!

Comment (yashykt) [inline test/core/end2end/tests/payload_with_proxy_auth.c]: done!

File: test/core/end2end/tests/payload_with_proxy_auth.c (hunk 24 above)
Comment (yashykt) [inline test/core/end2end/tests/payload_with_proxy_auth.c]: Made changes to generate_tests.bzl and gen_build_yaml.py

File: src/core/ext/filters/client_channel/http_proxy.c [hunk 31]
```diff
@@ -58,27 +56,25 @@ static char *grpc_get_http_proxy_server(grpc_exec_ctx* exec_ctx,
...
-    }
-    proxy_name = gpr_strdup(authority_strs[1]);
+  if (authority_nstrs == 1) {
+    /* User cred not present in authority */
+    proxy_name = authority_strs[0];
+  } else if (authority_nstrs == 2) {
+    /* User cred found */
+    *user_cred = authority_strs[0];
+    proxy_name = authority_strs[1];
     gpr_log(GPR_INFO, "userinfo found in proxy URI");
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/http_proxy.c]: I suggest changing this to GPR_DEBUG.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 32]
```diff
@@ -309,41 +309,23 @@ static void on_server_connect_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+ * Basic <base64_encoded_expected_cred>
  * Returns true if it matches, false otherwise
  */
-static bool proxy_auth_header_matches(grpc_exec_ctx *exec_ctx,
-                                      char *proxy_auth_header_val) {
-  if(proxy_auth_header_val == NULL) {
+static bool proxy_auth_header_matches(grpc_exec_ctx* exec_ctx,
+                                      char* proxy_auth_header_val,
+                                      char* expected_cred) {
+  GPR_ASSERT(proxy_auth_header_val != NULL && expected_cred != NULL);
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Let's split this into two separate asserts, so that if they fire, we'll be able to tell which one is the problem.

File: test/core/end2end/fixtures/h2_http_proxy.c [hunk 33]
```diff
@@ -79,13 +68,13 @@ void chttp2_init_client_fullstack(grpc_end2end_test_fixture *f,
   char *proxy_uri;
 
   /* If testing for proxy auth, add credentials to proxy uri */
-  if(grpc_channel_args_find(
-      client_args, GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT) == NULL) {
+  const grpc_arg *proxy_auth =
```
Comment (markdroth) [inline test/core/end2end/fixtures/h2_http_proxy.c]: Suggest calling this `proxy_auth_arg`.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 34]
```diff
@@ -395,23 +377,21 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
     return;
   }
   // If proxy auth is being used, check if the header is present and as expected
-  if(grpc_channel_args_find(
-      conn->proxy->channel_args,
-      GRPC_END2END_HTTP_PROXY_TEST_CONNECT_AUTH_PRESENT) != NULL) {
+  const grpc_arg* proxy_auth = grpc_channel_args_find(
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Suggest calling this `proxy_auth_arg`.

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 35]
```diff
@@ -395,23 +377,21 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
...
+      conn->proxy->channel_args, GRPC_ARG_HTTP_PROXY_AUTH_CREDS);
+  if (proxy_auth != NULL) {
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: `&& proxy_auth->type == GRPC_ARG_STRING`

File: test/core/end2end/fixtures/h2_http_proxy.c [hunk 36]
```diff
@@ -79,13 +68,13 @@ void chttp2_init_client_fullstack(grpc_end2end_test_fixture *f,
...
+      grpc_channel_args_find(client_args, GRPC_ARG_HTTP_PROXY_AUTH_CREDS);
+  if (proxy_auth == NULL) {
```
Comment (markdroth) [inline test/core/end2end/fixtures/h2_http_proxy.c]: `|| proxy_auth->type != GRPC_ARG_STRING`

File: test/core/end2end/fixtures/http_proxy_fixture.c [hunk 37]
```diff
@@ -395,23 +377,21 @@ static void on_read_request_done(grpc_exec_ctx* exec_ctx, void* arg,
...
     bool auth_header_found = false;
```
Comment (markdroth) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: Suggest renaming this `client_authenticated`.

File: src/core/ext/filters/client_channel/http_proxy.c (hunk 31 above)
Comment (yashykt) [inline src/core/ext/filters/client_channel/http_proxy.c]: done!

File: test/core/end2end/fixtures/h2_http_proxy.c (hunk 36 above)
Comment (yashykt) [inline test/core/end2end/fixtures/h2_http_proxy.c]: done!

File: test/core/end2end/fixtures/h2_http_proxy.c (hunk 33 above)
Comment (yashykt) [inline test/core/end2end/fixtures/h2_http_proxy.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 32 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 34 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 35 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

File: test/core/end2end/fixtures/http_proxy_fixture.c (hunk 37 above)
Comment (yashykt) [inline test/core/end2end/fixtures/http_proxy_fixture.c]: done!

**Comment by dgquintas** [comment]
Issues: #11188 #11737

**Comment by yashykt** [comment]
Thanks for reviewing @dgquintas @markdroth
